### PR TITLE
Add Emacs 23 in builds

### LINKED ISSUE #10: the Dockerfiles in those subdirs aren't actually used
The way `functions.sh` is written only the top level `Dockerfile` is used in the builds. Version-specific  scripts like `25.1/Dockerfile` are not consulted.  It doesn't matter right now because the Dockerfiles are currently identical, but I have a pull request underway for Emacs 23 where the base image and the build steps need to be different. So I need a way to separate them.

My plan is to rewrite `functions.sh` to launch `docker build -f Dockerfile.$tag` and then customize `Dockerfile.23` in the way I need.

### PR DESCRIPTION
**Pull request by gonewest818**
Due to the age of Emacs 23, I found it needs different steps to bootstrap the build, and it was cleaner to build it on Ubuntu 12.04. So therefore, this task required a reorganization of the bash script so that version specific Dockerfiles were used [Fixes #10]. Now what you see is three different build configurations in `dockerfiles/` to which we create version specific symlinks at the top level directory.  In other words the directory structure was changed like this:

screen shot 2018-02-20 at 10 06 44 am

I also noticed the builds are cloning the git repo even for tagged releases for which source tarballs are available. That seemed like an expensive operation and we're never going to use the git history in the build. I changed the Dockerfile to fetch just the source tarball for numbered releases rather than clone the repo. Now only the "master" branch build clones from git.

### CODE REVIEW

File: functions.sh [hunk 1]
```diff
@@ -1,18 +1,31 @@
...
+DOCKER_REPO=$(echo $TRAVIS_REPO_SLUG | cut -d "/" -f 1 | tr '[:upper:]' '[:lower:]')
+DOCKER_IMG=emacs
+
 build()
 {
-  for tag in $DOCKER_TAGS; do
-    docker build -t silex/emacs:$tag --build-arg="GIT_BRANCH=$GIT_BRANCH" .
-  done
+    for tag in $DOCKER_TAGS; do
+        major=$(echo $tag | cut -d "." -f 1)  # i.e. "24.5" => "25"
```
Comment (Silex) [inline functions.sh]: The comment is wrong here no? It'd be `24.5 => 24`

File: functions.sh [hunk 2]
```diff
@@ -1,18 +1,31 @@
...
-      docker push silex/emacs:$tag
-    done
-  fi
+    if [ "$TRAVIS_BRANCH" == "master" ]; then
+        docker login -u="$DOCKER_USERNAME" -p="$DOCKER_PASSWORD"
+        for tag in $DOCKER_TAGS; do
+            echo Push $DOCKER_REPO/$DOCKER_IMG:$tag
+            docker push $DOCKER_REPO/$DOCKER_IMG:$tag
+        done
+    fi
```
Comment (Silex) [inline functions.sh]: Looks like you use intendation at 4 spaces. This project used 2 spaces until now...

File: dockerfiles/Dockerfile.ubuntu12.from-tar [hunk 3]
```diff
@@ -24,14 +24,21 @@ RUN apt-get update && \
...
 ARG GIT_BRANCH
-RUN git clone --depth 1 --branch $GIT_BRANCH $GIT_REPOSITORY /tmp/emacs && \
-    cd /tmp/emacs && \
-    ./autogen.sh && \
-    ./configure && \
+
+# Tagged releases are available as tar balls, and should be smaller
+# and faster to download.  The build steps below are specifically for
+# Emacs 23, not correct for 24+.
+
```
Comment (Silex) [inline dockerfiles/Dockerfile.ubuntu12.from-tar]: Hum, if we use the tagged release over HTTP then we'd rename the `GIT_*` variables too.
I'm not sure we really care about the source clone speed tho, this is all done in travis-ci and the clone/download time is likely small compared to the emacs building time.

That said a bit of optimisation does not hurt. Are the tarballs always with the latest patches? The emacs release scheme always confuses me :confused:

**Comment by Silex** [comment]
Thanks! I'm generally ok with the direction this is going but will probably nitpick there and there about a few things :-)

**Comment by gonewest818** [comment]
np, I'll take care of the changes.

**Comment by Silex** [comment]
Alright, will check this tomorrow (going to bed now). Thanks!

File: dockerfiles/Dockerfile.ubuntu12.from-tar (hunk 3 above)
Comment (gonewest818) [reply dockerfiles/Dockerfile.ubuntu12.from-tar]: At one point I found Emacs developer docs stating their release numbering conventions. I can't find it right now, but ....

Only the major and minor number is considered meaningful ("25.3").  If there is a third "build" number (as in "26.0.91") then that indicates a development build and not a true release. There are ad hoc conventions how they use the third number in their development builds but there's no guarantee those conventions will be strictly followed.

If there is a major flaw or security issue then what I see them do is release a new version (25.3 was such a case, a CVE was reported on 25.2 and earlier). You can see the timeline of tags in their git repo is strictly ordered too.

Comment (gonewest818) [reply dockerfiles/Dockerfile.ubuntu12.from-tar]: About using the tar files and maybe changing the `GIT_*` variables, let me do some more experimentation on the sequential builds... if I'm asking Travis CI to cache the source trees anyway then this optimization isn't really necessary.

**Comment by gonewest818** [comment]
This last commit should address all your feedback but Travis CI jobs are stalling and timing out(*) right in the middle of cloning the emacs.git repo. I can't see if my changes are working or not. I'll have to step away and come back later.

(*) a suitable reminder why I'm doing this work in the first place, I'm building these docker images to run unit tests on emacs packages in CircleCI.

**Comment by Silex** [comment]
Thanks :+1:

**Pull request merged**